ios/Runner/AppDelegate.swift

```diff
@@ -3,11 +3,72 @@ import Flutter
 
 @main
 @objc class AppDelegate: FlutterAppDelegate {
+  private var notificationTapEventListener: NotificationTapEventListener?
+
   override func application(
     _ application: UIApplication,
     didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
   ) -> Bool {
     GeneratedPluginRegistrant.register(with: self)
+    guard let controller = window?.rootViewController as? FlutterViewController else {
+      fatalError("rootViewController is not type FlutterViewController")
+    }
+
+    // Retrieve the remote notification payload from launch options;
+    // this will be null if the launch wasn't triggered by a notification.
+    let notificationPayload = launchOptions?[.remoteNotification] as? [AnyHashable : Any]
+    let api = NotificationHostApiImpl(notificationPayload.map { NotificationDataFromLaunch(payload: $0) })
+    NotificationHostApiSetup.setUp(binaryMessenger: controller.binaryMessenger, api: api)
+
+    notificationTapEventListener = NotificationTapEventListener()
+    NotificationTapEventsStreamHandler.register(with: controller.binaryMessenger, streamHandler: notificationTapEventListener!)
+
+    // Setup handler for notification tap while the app is running.
+    UNUserNotificationCenter.current().delegate = self
+
     return super.application(application, didFinishLaunchingWithOptions: launchOptions)
   }
+
+  override func userNotificationCenter(
+    _ center: UNUserNotificationCenter,
+    didReceive response: UNNotificationResponse,
+    withCompletionHandler completionHandler: @escaping () -> Void
+  ) {
+    if let listener = notificationTapEventListener {
+      let userInfo = response.notification.request.content.userInfo
+      listener.onNotificationTapEvent(event: NotificationTapEvent(payload: userInfo))
+      completionHandler()
+    }
+  }
+}
+
+private class NotificationHostApiImpl: NotificationHostApi {
+  private let maybeDataFromLaunch: NotificationDataFromLaunch?
+
+  init(_ maybeDataFromLaunch: NotificationDataFromLaunch?) {
+    self.maybeDataFromLaunch = maybeDataFromLaunch
+  }
+
+  func getNotificationDataFromLaunch() -> NotificationDataFromLaunch? {
+    maybeDataFromLaunch
+  }
+}
+
+class NotificationTapEventListener: NotificationTapEventsStreamHandler {
+  var eventSink: PigeonEventSink<NotificationTapEvent>?
+
+  override func onListen(withArguments arguments: Any?, sink: PigeonEventSink<NotificationTapEvent>) {
+    eventSink = sink
+  }
+
+  func onNotificationTapEvent(event: NotificationTapEvent) {
+    if let eventSink = eventSink {
+      eventSink.success(event)
+    }
```

Comment:
This would do the same thing, right?
```suggestion
    eventSink?.success(event)
```